File: src/main/java/nu/fgv/register/server/spex/SpexCategoryService.java
package nu.fgv.register.server.spex;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nu.fgv.register.server.util.FileUtil;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static nu.fgv.register.server.spex.SpexCategoryMapper.SPEX_CATEGORY_MAPPER;
import static org.springframework.util.StringUtils.hasText;

@Slf4j
@RequiredArgsConstructor
@Service
@Transactional
public class SpexCategoryService {

    private final SpexCategoryRepository repository;

    public List<SpexCategoryDto> findAll(final Sort sort) {
        return repository
                .findAll(sort)
                .stream()
                .map(SPEX_CATEGORY_MAPPER::toDto)
                .collect(Collectors.toList());
    }

    public Page<SpexCategoryDto> find(final Pageable pageable) {
        return repository
                .findAll(pageable)
                .map(SPEX_CATEGORY_MAPPER::toDto);
    }

    public Optional<SpexCategoryDto> findById(final Long id) {
        return repository
                .findById(id)
                .map(SPEX_CATEGORY_MAPPER::toDto);
    }

    public List<SpexCategoryDto> findByIds(final List<Long> ids, final Sort sort) {
        return repository
                .findByIds(ids, sort)
                .stream()
                .map(SPEX_CATEGORY_MAPPER::toDto)
                .collect(Collectors.toList());
    }

    public SpexCategoryDto create(final SpexCategoryCreateDto dto) {
        return SPEX_CATEGORY_MAPPER.toDto(repository.save(SPEX_CATEGORY_MAPPER.toModel(dto)));
    }

    public Optional<SpexCategoryDto> update(final SpexCategoryUpdateDto dto) {
        return repository
                .findById(dto.getId())
                .map(category -> SPEX_CATEGORY_MAPPER.toModel(dto))
                .map(repository::save)
                .map(SPEX_CATEGORY_MAPPER::toDto);
    }

    public Optional<SpexCategoryDto> partialUpdate(final SpexCategoryUpdateDto dto) {
        return repository
                .findById(dto.getId())
                .map(model -> {
                    SPEX_CATEGORY_MAPPER.toPartialModel(dto, model);
                    return model;
                })
                .map(repository::save)
                .map(SPEX_CATEGORY_MAPPER::toDto);
    }

    public void deleteById(final Long id) {
        repository.deleteById(id);
    }

    public Optional<SpexCategoryDto> saveLogo(final Long id, final byte[] logo, final String contentType) {
        return repository
                .findById(id)
                .map(model -> {
                    model.setLogo(logo);
                    model.setLogoContentType(hasText(contentType) ? contentType : FileUtil.detectMimeType(logo));
                    repository.save(model);
                    return SPEX_CATEGORY_MAPPER.toDto(model);
                });
    }

    public Optional<SpexCategoryDto> removeLogo(final Long id) {
        return repository
                .findById(id)
                .map(model -> {
                    model.setLogo(null);
                    model.setLogoContentType(null);
                    repository.save(model);
                    return SPEX_CATEGORY_MAPPER.toDto(model);
                });
    }

    public Optional<Pair<byte[], String>> getLogo(final Long id) {
        return repository.findById(id)
                .filter(model -> model.getLogo() != null && hasText(model.getLogoContentType()))
                .map(model -> Pair.of(model.getLogo(), model.getLogoContentType()));
    }
}

File: src/main/kotlin/nu/fgv/register/server/service/SpexCategoryService.kt
package nu.fgv.register.server.service

import java.util.*
import nu.fgv.register.server.dto.SpexCategoryDto
import nu.fgv.register.server.dto.mapper.SpexCategoryMapper
import nu.fgv.register.server.repository.SpexCategoryRepository
import nu.fgv.register.server.repository.search.SpexCategorySearchRepository
import org.elasticsearch.index.query.QueryBuilders.queryStringQuery
import org.slf4j.LoggerFactory.getLogger
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class SpexCategoryService(
    private val repository: SpexCategoryRepository,
    private val mapper: SpexCategoryMapper,
    private val searchRepository: SpexCategorySearchRepository
) {

    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        @JvmStatic
        private val log = getLogger(javaClass.enclosingClass)
    }

    fun save(spexCategoryDto: SpexCategoryDto): SpexCategoryDto {
        log.debug("Request to save spex category : {}") { spexCategoryDto }

        var spexCategory = mapper.toEntity(spexCategoryDto)
        val result = repository.save(spexCategory)
        searchRepository.save(result)
        return mapper.toDto(result)
    }

    @Transactional(readOnly = true)
    fun findAll(pageable: Pageable): Page<SpexCategoryDto> {
        log.debug("Request to get all spex categories")
        return repository.findAll(pageable).map(mapper::toDto)
    }

    @Transactional(readOnly = true)
    fun findOne(id: Long): Optional<SpexCategoryDto> {
        log.debug("Request to get spex category : {}") { id }
        return repository.findById(id).map(mapper::toDto)
    }

    fun delete(id: Long) {
        log.debug("Request to delete spex category : {}") { id }
        repository.deleteById(id)
        searchRepository.deleteById(id)
    }

    @Transactional(readOnly = true)
    fun search(query: String, pageable: Pageable): Page<SpexCategoryDto>? {
        log.debug("Request to search for a page of spex categories for query {}") { query }
        return searchRepository.search(queryStringQuery(query), pageable).map(mapper::toDto)
    }
}
